### DISCUSSION TITLE
Excessive Memory fragmentation

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#15153] [RP2040 / Pico]
I am having a lot of trouble with MemoryError Exceptions being thrown on the PICO_W due to fragmentation for small requests, like a couple of hundred bytes. Especially in the logger. These memory requirements are so small there must be something I am doing wrong to lead to such fragmentations. If an operation (often in the base micropython - not in my code ) can't even come up with 200+ bytes to do some operation it's going to be next to impossible to get something done.

What could be causing such fragmentations? Is it the logger? Creation of strings? I would like to be able to pre-allocate a fixed buffer to handle the messages in the logger but in Python strings are immutable. It would be great if one could populate the fixed memory as a bytearray and then have python handle it as if it were a string. But going from a bytearray to a string always involves a copy which requires even more memory.

In the end I think there must be something I am doing wrong to have such allocation issues on such small sizes as no one would be able to get anything done under those restrictions.

Would be grateful for any insight on how to deal with this or even would like to know if others have having similar issues.

**Comment by peterhinch** [comment]
It's hard to comment with so few details of the application but a starting point is to read [this official doc](https://docs.micropython.org/en/latest/reference/constrained.html).

**Reply by brianreinhold** [reply]
We have been looking at that quite a bit to see what we could do. The application connects to BLE health devices and there is a class for each BT-SIG profile, such as Pulse Oximeters, BP cuffs, weight scales, etc. Each time a connection to one of these devices is made, the class for that device type is instantiated. Devices from different companies will (should) follow the BT SIG profile, but they will differ, of course, on the amount of features they support. From what I have read it seems fragmentation is impossible to avoid if I make sequences of connections to even a given type of device (such as a Pulse Ox) but from different companies. Can one reserve space for these objects such that the same space is used for each connection (such as a memoryview)?
